File: compute_ground_truth.py
#!/usr/bin/env python3
"""
compute_ground_truth.py

读取数据集目录下的 base.txt（每行是向量，可能以 id 开头），随机选取若干 query（流式水塘抽样），
对每个 query 对整个数据集做流式遍历计算距离（L2 或 cosine），维护 top-k 最近邻（最小距离），
并把结果写成 JSON。

可选参数：--dataset 指定单个文件，未指定则尝试处理 data_o/glove/base.txt 和 data_o/sift/base.txt。
支持 --num_queries, --k, --metric (l2|cosine), --seed, --out, --max_base (仅处理前 N 个向量，便于 smoke-test)

实现目标：对大文件也能工作（不把整个文件一次性读入内存）。
"""

from __future__ import annotations

import argparse
import math
import heapq
import json
import random
from typing import List, Optional, Tuple, Iterable
import threading
from queue import Queue


def parse_vector_line(line: str) -> Optional[Tuple[Optional[str], List[float]]]:
    s = line.strip()
    if not s:
        return None
    parts = s.split()
    # try to convert all to float
    try:
        vals = [float(x) for x in parts]
        # no id
        return None, vals
    except ValueError:
        # assume first token is id, rest are floats
        if len(parts) < 2:
            return None
        id0 = parts[0]
        try:
            vals = [float(x) for x in parts[1:]]
            return id0, vals
        except ValueError:
            return None


def l2_distance(a: List[float], b: List[float]) -> float:
    # assume same length
    s = 0.0
    for x, y in zip(a, b):
        d = x - y
        s += d * d
    return math.sqrt(s)


def cosine_distance(a: List[float], b: List[float]) -> float:
    dot = 0.0
    na = 0.0
    nb = 0.0
    for x, y in zip(a, b):
        dot += x * y
        na += x * x
        nb += y * y
    if na == 0 or nb == 0:
        return 1.0
    cosine_sim = dot / (math.sqrt(na) * math.sqrt(nb))
    # convert similarity to distance in [0,2]
    return 1.0 - cosine_sim


def reservoir_sample_vectors(path: str, num: int, seed: int = 1) -> List[Tuple[int, Optional[str], List[float]]]:
    """流式从文件中随机抽取 num 个向量，返回 (line_index, id_or_None, vector) 的列表"""
    random.seed(seed)
    reservoir: List[Tuple[int, Optional[str], List[float]]] = []
    with open(path, "r", encoding="utf-8", errors="ignore") as f:
        for idx, line in enumerate(f):
            parsed = parse_vector_line(line)
            if parsed is None:
                continue
            id0, vec = parsed
            item = (idx, id0, vec)
            if len(reservoir) < num:
                reservoir.append(item)
            else:
                j = random.randrange(idx + 1)
                if j < num:
                    reservoir[j] = item
    return reservoir


def topk_bruteforce_stream(path: str, queries: List[Tuple[int, Optional[str], List[float]]], k: int, metric: str = "l2", max_base: Optional[int] = None, threads: int = 4, batch_size: int = 1000, show_progress: bool = True):
    """多线程生产者-消费者实现的流式暴力 top-k 搜索。
    主线程分批读取文件（每 batch_size 行为一块），放入队列；工作线程消费块并更新每个 query 的堆。
    """
    metric_fn = l2_distance if metric == "l2" else cosine_distance

    # 每个 query 有自己的堆和锁
    heaps: List[List[Tuple[float, int, Optional[str], float]]] = [ [] for _ in range(len(queries)) ]
    locks = [threading.Lock() for _ in queries]

    q = Queue(maxsize=threads * 2)
    stop_sentinels = threads

    # try to import tqdm for nicer progress bar
    try:
        from tqdm import tqdm
        use_tqdm = True
    except Exception:
        tqdm = None
        use_tqdm = False

    processed = 0
    processed_lock = threading.Lock()

    def worker():
        nonlocal processed
        while True:
            batch = q.get()
            if batch is None:
                q.task_done()
                break
            for idx, id0, vec in batch:
                for qi, (_, qid, qvec) in enumerate(queries):
                    try:
                        dist = metric_fn(qvec, vec)
                    except Exception:
                        dist = float('inf')
                    # update heap for query qi
                    with locks[qi]:
                        if len(heaps[qi]) < k:
                            heapq.heappush(heaps[qi], (-dist, idx, id0, dist))
                        else:
                            if -heaps[qi][0][0] > dist:
                                heapq.heapreplace(heaps[qi], (-dist, idx, id0, dist))
                with processed_lock:
                    processed += 1
            q.task_done()

    # start worker threads
    workers = []
    for _ in range(max(1, threads)):
        t = threading.Thread(target=worker, daemon=True)
        t.start()
        workers.append(t)

    # Reading producer
    total_estimate = max_base if max_base is not None else None
    pbar = None
    if show_progress:
        if use_tqdm:
            pbar = tqdm(total=total_estimate, unit='vec')

    try:
        with open(path, 'r', encoding='utf-8', errors='ignore') as f:
            batch = []
            for idx, line in enumerate(f):
                if max_base is not None and idx >= max_base:
                    break
                parsed = parse_vector_line(line)
                if parsed is None:
                    continue
                id0, vec = parsed
                batch.append((idx, id0, vec))
                if len(batch) >= batch_size:
                    q.put(batch)
                    if pbar is not None:
                        pbar.update(len(batch))
                    batch = []
            if batch:
                q.put(batch)
                if pbar is not None:
                    pbar.update(len(batch))
    finally:
        # send stop sentinels
        for _ in range(stop_sentinels):
            q.put(None)
        # wait for queue to be fully processed
        q.join()
        if pbar is not None:
            pbar.close()

    # collect results
    results = []
    for qi, (qidx, qid, qvec) in enumerate(queries):
        h = heaps[qi]
        neighbors = sorted([{'index': item[1], 'id': item[2], 'distance': item[3]} for item in h], key=lambda x: x['distance'])
        results.append({'query_index': qidx, 'query_id': qid, 'query_vector_len': len(qvec), 'neighbors': neighbors})
    return results


def process_dataset(path: str, num_queries: int, k: int, metric: str, seed: int, max_base: Optional[int], out: Optional[str], threads: int = 4, batch_size: int = 1000, show_progress: bool = True):
    print(f"Processing {path} (num_queries={num_queries}, k={k}, metric={metric}, max_base={max_base}, threads={threads}, batch_size={batch_size})")
    queries = reservoir_sample_vectors(path, num_queries, seed=seed)
    if not queries:
        print("No valid vectors found in", path)
        return None
    results = topk_bruteforce_stream(path, queries, k, metric=metric, max_base=max_base, threads=threads, batch_size=batch_size, show_progress=show_progress)
    payload = {
        'dataset': path,
        'metric': metric,
        'k': k,
        'num_queries': num_queries,
        'results': results,
    }
    if out:
        with open(out, 'w', encoding='utf-8') as fo:
            json.dump(payload, fo, ensure_ascii=False, indent=2)
        print(f"Wrote output to {out}")
    else:
        print(json.dumps(payload, ensure_ascii=False, indent=2))
    return payload


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--dataset', type=str, help='单个 base.txt 文件路径')
    parser.add_argument('--num_queries', type=int, default=1000)
    parser.add_argument('--k', type=int, default=10)
    parser.add_argument('--metric', type=str, choices=['l2', 'cosine'], default='l2')
    parser.add_argument('--seed', type=int, default=1)
    parser.add_argument('--out', type=str, help='输出 JSON 文件路径（如果不指定则打印）')
    parser.add_argument('--max_base', type=int, default=None, help='仅处理前 N 个基向量（用于快速 smoke-test）')
    parser.add_argument('--threads', type=int, default=32, help='工作线程数')
    parser.add_argument('--batch_size', type=int, default=1000, help='主线程每次读入的向量批大小')
    parser.add_argument('--no_progress', action='store_true', help='禁用进度条（默认显示）')
    args = parser.parse_args()

    datasets = []
    if args.dataset:
        datasets = [args.dataset]
    else:
        datasets = [
            'data_o/glove/base.txt',
            'data_o/sift/base.txt'
        ]

    for ds in datasets:
        out_path = None
        if args.out:
            # if multiple datasets and single out specified, append dataset name
            if len(datasets) > 1:
                out_path = args.out.rsplit('.', 1)[0] + '_' + ds.replace('/', '_').replace('\\', '_').replace(':','') + '.json'
            else:
                out_path = args.out
        else:
            out_path = None
        try:
            process_dataset(ds, args.num_queries, args.k, args.metric, args.seed, args.max_base, out_path, threads=args.threads, batch_size=args.batch_size, show_progress=not args.no_progress)
        except FileNotFoundError:
            print(f"File not found: {ds}")
        except Exception as e:
            print(f"Error processing {ds}: {e}")


if __name__ == '__main__':
    main()

File: compute_ground_truth.cpp
// compute_ground_truth.cpp
// 多线程生产者-消费者暴力 top-k 最近邻 (流式、可限制扫描条数)
// 支持：reservoir 抽样 query，--num_queries, --k, --metric (l2|cosine), --threads, --batch_size, --max_base, --out
// 输出：简单 JSON 文件（无外部依赖）

#include <algorithm>
#include <atomic>
#include <condition_variable>
#include <cctype>
#include <fstream>
#include <functional>
#include <iostream>
#include <limits>
#include <mutex>
#include <numeric>
#include <queue>
#include <random>
#include <sstream>
#include <string>
#include <thread>
#include <unordered_map>
#include <utility>
#include <vector>

using namespace std;

struct VecItem {
    int index;
    string id; // empty if none
    vector<double> v;
};

bool try_stod(const string &s, double &out) {
    try {
        size_t pos;
        out = stod(s, &pos);
        return pos == s.size();
    } catch (...) {
        return false;
    }
}

bool parse_vector_line(const string &line, string &out_id, vector<double> &out_vec) {
    out_id.clear();
    out_vec.clear();
    istringstream iss(line);
    vector<string> toks;
    string t;
    while (iss >> t) toks.push_back(t);
    if (toks.empty()) return false;
    double val;
    // try treat ALL tokens as numbers
    bool allnum = true;
    for (auto &s : toks) {
        if (!try_stod(s, val)) { allnum = false; break; }
    }
    if (allnum) {
        out_id.clear();
        out_vec.reserve(toks.size());
        for (auto &s : toks) out_vec.push_back(stod(s));
        return true;
    }
    // otherwise first token is id
    if (toks.size() < 2) return false;
    out_id = toks[0];
    out_vec.reserve(toks.size()-1);
    for (size_t i = 1; i < toks.size(); ++i) {
        if (!try_stod(toks[i], val)) return false;
        out_vec.push_back(stod(toks[i]));
    }
    return true;
}

double l2_distance(const vector<double> &a, const vector<double> &b) {
    double s = 0.0;
    size_t n = min(a.size(), b.size());
    for (size_t i = 0; i < n; ++i) {
        double d = a[i] - b[i]; s += d*d;
    }
    return sqrt(s);
}

double cosine_distance(const vector<double> &a, const vector<double> &b) {
    double dot = 0.0, na = 0.0, nb = 0.0;
    size_t n = min(a.size(), b.size());
    for (size_t i = 0; i < n; ++i) {
        dot += a[i]*b[i]; na += a[i]*a[i]; nb += b[i]*b[i];
    }
    if (na == 0.0 || nb == 0.0) return 1.0;
    double sim = dot / (sqrt(na)*sqrt(nb));
    return 1.0 - sim;
}

// Reservoir sampling queries from file
vector<VecItem> reservoir_sample(const string &path, int num, unsigned seed) {
    vector<VecItem> reservoir;
    reservoir.reserve(num);
    ifstream ifs(path);
    if (!ifs) return reservoir;
    string line;
    std::mt19937 rng(seed);
    int idx = 0;
    while (std::getline(ifs, line)) {
        string id; vector<double> vec;
        if (!parse_vector_line(line, id, vec)) { ++idx; continue; }
        VecItem it{idx, id, std::move(vec)};
        if ((int)reservoir.size() < num) {
            reservoir.push_back(std::move(it));
        } else {
            uniform_int_distribution<int> dist(0, idx);
            int j = dist(rng);
            if (j < num) reservoir[j] = std::move(it);
        }
        ++idx;
    }
    return reservoir;
}

// Thread-safe queue of batches
template<typename T>
class ThreadQueue {
public:
    void push(T &&v) {
        {
            unique_lock<mutex> lk(m_);
            q_.push(std::move(v));
        }
        cv_.notify_one();
    }
    bool pop(T &out) {
        unique_lock<mutex> lk(m_);
        cv_.wait(lk, [&]{ return !q_.empty() || finished_; });
        if (q_.empty()) return false;
        out = std::move(q_.front()); q_.pop();
        return true;
    }
    void set_finished() {
        {
            unique_lock<mutex> lk(m_);
            finished_ = true;
        }
        cv_.notify_all();
    }
private:
    queue<T> q_;
    mutex m_;
    condition_variable cv_;
    bool finished_ = false;
};

struct QueryState {
    int qidx;
    string id;
    vector<double> vec;
    // max-heap of size up to k; store pair(distance, base_index). largest distance on top
    priority_queue<pair<double,int>> heap;
    mutex m;
};

void worker_func(ThreadQueue<vector<VecItem>> &tq, vector<unique_ptr<QueryState>> &queries, int k, const string &metric, atomic<long long> &processed) {
    vector<VecItem> batch;
    while (tq.pop(batch)) {
        for (auto &b : batch) {
            for (auto &qptr : queries) {
                auto &q = *qptr;
                double dist = numeric_limits<double>::infinity();
                if (metric == "l2") dist = l2_distance(q.vec, b.v);
                else dist = cosine_distance(q.vec, b.v);
                // update q.heap
                {
                    lock_guard<mutex> lk(q.m);
                    if ((int)q.heap.size() < k) q.heap.emplace(dist, b.index);
                    else if (q.heap.top().first > dist) {
                        q.heap.pop(); q.heap.emplace(dist, b.index);
                    }
                }
            }
            ++processed;
        }
    }
}

void write_json_output(const string &out_path, const string &dataset, const vector<unique_ptr<QueryState>> &queries, const string &metric, int k) {
    ofstream ofs(out_path);
    if (!ofs) {
        cerr << "Cannot open output file: " << out_path << "\n";
        return;
    }
    ofs << "{\n";
    ofs << "  \"dataset\": \"" << dataset << "\",\n";
    ofs << "  \"metric\": \"" << metric << "\",\n";
    ofs << "  \"k\": " << k << ",\n";
    ofs << "  \"results\": [\n";
    for (size_t i = 0; i < queries.size(); ++i) {
        const auto &q = *queries[i];
        // extract neighbors sorted ascending by distance
        vector<pair<double,int>> list;
        {
            // copy heap
            auto tmp = q.heap;
            while (!tmp.empty()) { list.push_back(tmp.top()); tmp.pop(); }
        }
        sort(list.begin(), list.end(), [](const auto &a, const auto &b){ return a.first < b.first; });
        ofs << "    {\n";
        ofs << "      \"query_index\": " << q.qidx << ",\n";
        ofs << "      \"query_id\": \"" << q.id << "\",\n";
        ofs << "      \"neighbors\": [\n";
        for (size_t j = 0; j < list.size(); ++j) {
            ofs << "        { \"index\": " << list[j].second << ", \"distance\": " << list[j].first << " }";
            if (j+1 < list.size()) ofs << ",";
            ofs << "\n";
        }
        ofs << "      ]\n";
        ofs << "    }";
        if (i+1 < queries.size()) ofs << ",";
        ofs << "\n";
    }
    ofs << "  ]\n";
    ofs << "}\n";
}

int main(int argc, char **argv) {
    // simple arg parsing
    string dataset = "";
    int num_queries = 100;
    int k = 5;
    string metric = "l2";
    int threads = 32;
    int batch_size = 1000;
    int max_base = -1;
    unsigned seed = 1;
    string out = "";

    for (int i = 1; i < argc; ++i) {
        string a = argv[i];
        if (a == "--dataset" && i+1 < argc) { dataset = argv[++i]; }
        else if (a == "--num_queries" && i+1 < argc) num_queries = stoi(argv[++i]);
        else if (a == "--k" && i+1 < argc) k = stoi(argv[++i]);
        else if (a == "--metric" && i+1 < argc) metric = argv[++i];
        else if (a == "--threads" && i+1 < argc) threads = stoi(argv[++i]);
        else if (a == "--batch_size" && i+1 < argc) batch_size = stoi(argv[++i]);
        else if (a == "--max_base" && i+1 < argc) max_base = stoi(argv[++i]);
        else if (a == "--seed" && i+1 < argc) seed = stoi(argv[++i]);
        else if (a == "--out" && i+1 < argc) out = argv[++i];
        else if (a == "--help") {
            cout << "Usage: compute_ground_truth.exe --dataset PATH [--num_queries N] [--k K] [--metric l2|cosine] [--threads T] [--batch_size B] [--max_base N] [--out out.json]\n";
            return 0;
        }
    }
    if (dataset.empty()) {
        cerr << "Please provide --dataset path\n";
        return 1;
    }

    cout << "Reservoir sampling " << num_queries << " queries from " << dataset << "...\n";
    auto queries_items = reservoir_sample(dataset, num_queries, seed);
    if (queries_items.empty()) { cerr << "No queries sampled (file empty or parse failed)\n"; return 1; }

    vector<unique_ptr<QueryState>> queries;
    for (size_t i = 0; i < queries_items.size(); ++i) {
        auto qs = std::make_unique<QueryState>();
        qs->qidx = queries_items[i].index;
        qs->id = std::move(queries_items[i].id);
        qs->vec = std::move(queries_items[i].v);
        queries.push_back(std::move(qs));
    }

    ThreadQueue<vector<VecItem>> tq;
    atomic<long long> processed{0};

    // start workers
    vector<thread> workers;
    for (int i = 0; i < threads; ++i)
        workers.emplace_back(worker_func, std::ref(tq), std::ref(queries), k, metric, std::ref(processed));

    // producer: read dataset in batches and push into queue
    ifstream ifs(dataset);
    if (!ifs) { cerr << "Cannot open dataset: " << dataset << "\n"; tq.set_finished(); for (auto &t: workers) t.join(); return 1; }
    string line; vector<VecItem> batch; batch.reserve(batch_size);
    int idx = 0;
    while (std::getline(ifs, line)) {
        if (max_base >= 0 && idx >= max_base) break;
        string id; vector<double> vec;
        if (!parse_vector_line(line, id, vec)) { ++idx; continue; }
        batch.push_back(VecItem{idx, id, std::move(vec)});
        ++idx;
        if ((int)batch.size() >= batch_size) { tq.push(std::move(batch)); batch.clear(); batch.reserve(batch_size); }
    }
    if (!batch.empty()) { tq.push(std::move(batch)); }
    tq.set_finished();

    // simple progress display
    while (processed < idx) {
        cout << "Processed base vectors (approx): " << processed << "\r" << flush;
        this_thread::sleep_for(chrono::milliseconds(200));
    }

    // join workers
    for (auto &t : workers) if (t.joinable()) t.join();
    cout << "\nFinished.\n";

    // write output
    string outpath = out.empty() ? dataset.substr(0, dataset.find_last_of(".")) + ".json" : out;
    write_json_output(outpath, dataset, queries, metric, k);
    cout << "Wrote output to " << outpath << "\n";

    return 0;
}
